### PR TITLE
Global Styles: Update REST controller override method and backport changes from Core

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
Updates Global Styles REST controller method override to use post-type registration filter instead of initializing.

## Why?
This avoids extra work and prevents both controls from initializing, which are later merged, and the core class is discarded. 

See [`register_rest_route`](https://developer.wordpress.org/reference/functions/register_rest_route/) for more details (the `$override` argument is the key 😄).

## Testing Instructions
1. Set logger somewhere in `WP_REST_Global_Styles_Controller_Gutenberg`.
2. Open a post or page.
3. Confirm that test messages are logged.

### Testing Instructions for Keyboard
Same.

### CODE REVIEW

File: lib/rest-api.php
Comment (ramonjd) [inline lib/rest-api.php]: Oh, that's what you [meant](https://github.com/WordPress/gutenberg/pull/65071#issuecomment-2345298670) 🤦🏻 

I just assumed we were doing this somewhere - but clearly not! 🤭 

Thanks!

Comment (Mamaduka) [reply lib/rest-api.php]: No override was working; it was just doing double work. Using a post-registration filter should be preferred when possible.

[Here](https://github.com/WordPress/gutenberg/pull/65071#issuecomment-2345260471), I meant why we were doing full override instead of extending the existing class and your [answer](https://github.com/WordPress/gutenberg/pull/65071#issuecomment-2345298670) was satisfying :)

Comment (ramonjd) [reply lib/rest-api.php]: Oh, thanks for confirming. Good to know my powers of comprehension are not completely borked 🤣 

I smoked tested this PR only, but things are working as expected with no PHP errors. I'll give it a better test in the morning (AEST) if no-one gets to it first.

Comment (Mamaduka) [reply lib/rest-api.php]: Thank you, @ramonjd!

I see one PHP Unit test failure specific to PHP 8.2 env. Investigating that now.

**Review by spacedmonkey — APPROVED** [review]

**Comment by Mamaduka** [comment]
I cannot replicate PHP unit test failures locally, making it harder to debug.

@ramonjd, can you check if they're also passing for you locally?

**Comment by ramonjd** [comment]
> can you check if they're also passing for you locally?

Yeah, they're passing for me locally too. 🤔 I YOLOed a rebase push to see if the universe is just playing tricks on us.

However, the tests that appear to be failing are related to two, fairly recent features: background images and style variations. So that's something to start with. I'll see if I can reproduce locally....

**Comment by ramonjd** [comment]
I'm stumped. It's almost as if the tests are ignoring the two features, e.g., 

https://github.com/WordPress/gutenberg/blob/a965ed0de183a12bf919e3c66f290cf17f7aaf7e/lib/class-wp-rest-global-styles-controller-gutenberg.php#L335-L335


https://github.com/WordPress/gutenberg/blob/a965ed0de183a12bf919e3c66f290cf17f7aaf7e/lib/class-wp-rest-global-styles-controller-gutenberg.php#L711-L720

Which makes me think it's running the `WP_REST_Global_Styles_Controller` class from whichever Core version is running. 
But even if it were, it looks like the Core classes have the same methods, so scratch that.

It smells of some race condition, e.g., Core methods executing without knowledge of the test fixtures 🤷🏻 

Fooling around with the hook priority didn't lead me anywhere.

cc @aaronrobertshaw just in case he has some tricks up his sleeve

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for the ping 👍 

> cc @aaronrobertshaw just in case he has some tricks up his sleeve

No tricks per se. Just that I don't have a super deep understanding of everything under the hood with our REST API so maybe I can ask some dumb questions that help spark ideas on explanations/fixes.

> Yeah, they're passing for me locally too.

Same for me. All tests passes regardless of if I isolate the single test, the controller tests, multisite, or all PHP tests.

The error for `test_update_item_with_custom_block_style_variations` is reminiscent of some failures that were occurring when the section styles feature was refactored right before 6.6 beta/RC. 

I'm struggling to bring repressed traumatic memories back to the surface but IIRC the error was related to the post not being found to update or something. Gutenberg had some [duotone actions that triggered block style variations to be registered](https://github.com/WordPress/gutenberg/pull/62712#issuecomment-2185787727) and also likely ensured creation of the global styles CPT entry. 

This very likely could be a red herring, so feel free to ignore it. I tried removing the Gutenberg duotone actions locally and destroying wp-env but the test still passed locally 🤷 

> Which makes me think it's running the WP_REST_Global_Styles_Controller class from whichever Core version is running.
But even if it were, it looks like the Core classes have the same methods, so scratch that.

One difference is that the `WP_REST_Global_Styles_Controller` in Gutenberg is extending `WP_REST_Controller` whereas `WP_REST_Global_Styles_Controller` in core extends `WP_REST_Posts_Controller`. If this PR is now "replacing" the controller class rather than adding additional routes, should it be updated to extend the posts controller as per core's?

**Comment by ramonjd** [comment]
> If this PR is now "replacing" the controller class rather than adding additional routes, should it be updated to extend the posts controller as per core's?

🤦🏻 Great spotting. That totally slipped past the net.

**Comment by ramonjd** [comment]
I'm prepping a commit for this branch that brings over all the changes from the Core class. They should have been synced I believe. 

It's partly my fault after the Core patch: https://github.com/WordPress/wordpress-develop/commit/6fc019aa711308d6eaa1e672498b2a789a0c584c#diff-58982cfca0cd603cc3ace01a90ef7f243c7842eb8bf99291667a54f9d94ceecc

I think this PR is a good home for it since we're defining the post controller.

It'll be a single commit, so it can be rolled back.

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (ramonjd) [inline lib/class-wp-rest-global-styles-controller-gutenberg.php:360]: The `private_title_format` filter add/remove wasn't backported from https://github.com/WordPress/wordpress-develop/commit/edfc2b0d89eeff4c5ffdea2fea0946cd00b66bac

cc @youknowriad to see whether this was intentional

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (ramonjd) [inline lib/class-wp-rest-global-styles-controller-gutenberg.php:39]: Done in Core in https://github.com/WordPress/wordpress-develop/commit/6fc019aa711308d6eaa1e672498b2a789a0c584c as part of inheriting `WP_REST_Posts_Controller`

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (ramonjd) [inline lib/class-wp-rest-global-styles-controller-gutenberg.php:418]: Done in Core in https://github.com/WordPress/wordpress-develop/commit/6fc019aa711308d6eaa1e672498b2a789a0c584c as part of inheriting `WP_REST_Posts_Controller`

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (ramonjd) [inline lib/class-wp-rest-global-styles-controller-gutenberg.php:445]: Done in Core in https://github.com/WordPress/wordpress-develop/commit/6fc019aa711308d6eaa1e672498b2a789a0c584c as part of inheriting `WP_REST_Posts_Controller`

**Comment by ramonjd** [comment]
> I'm prepping a commit for this branch that brings over all the changes from the Core class. They should have been synced I believe.

Okay, that did absolutely nothing to fix the tests 🤣

**Comment by ramonjd** [comment]
Actually, and I missed this too, I think it's because an earlier [commit](https://github.com/WordPress/gutenberg/pull/65259/commits/457daea84a65ee31519af4e816150b53328d2ae0) removed the `register_routes()`

Will test that

**Comment by ramonjd** [comment]
> Actually, and I missed this too, I think it's because an earlier [commit](https://github.com/WordPress/gutenberg/pull/65259/commits/457daea84a65ee31519af4e816150b53328d2ae0) removed the register_routes()
> 
> Will test that


> Okay, that did absolutely nothing to fix the tests 🤣

Ditto.

**Comment by ramonjd** [comment]
Okay, sorry for all the noise @Mamaduka and @aaronrobertshaw  

I'm still stumped. I wonder if it's a sign of a deeper bug.

Wondering if we need to define the revisions controller too. See: https://github.com/WordPress/wordpress-develop/blob/6b81c17cb77ee1bd5fd672a689a1fec85b497161/src/wp-includes/post.php#L489

I did that, but I doubt that will affect the failing tests, which are the same failing tests as with previous commits 🤔

**Comment by ramonjd** [comment]
Okay!

I can confirm it's because the unit tests are running on WP 6.5.5

https://github.com/WordPress/gutenberg/actions/runs/10878265051/job/30180961587?pr=65259#step:13:75

I have reproduced this locally... now to debug

**Comment by ramonjd** [comment]
It's because we're missing

```php
	$args['show_in_rest']                    = true;
	$args['rest_base']                       = 'global-styles';
```

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (ramonjd) [inline lib/class-wp-rest-global-styles-controller-gutenberg.php:69]: This is required to overwrite Core routes when this endpoint is registered.

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (ramonjd) [inline lib/class-wp-rest-global-styles-controller-gutenberg.php:224]: Done in Core in https://github.com/WordPress/wordpress-develop/commit/6fc019aa711308d6eaa1e672498b2a789a0c584c as part of inheriting `WP_REST_Posts_Controller`

File: lib/rest-api.php
Comment (ramonjd) [inline lib/rest-api.php:27]: The CI tests are running 6.5.5. I think it's computed for CI [here](https://github.com/WordPress/gutenberg/blob/43d70f9af460406bab79c6a79bd4323a81e93766/.github/workflows/unit-test.yml#L95-L95).

They should probably be running 6.6, but even so, adding the args for backwards compat just in case.

Compare 6.5 https://github.com/WordPress/wordpress-develop/blob/6.5/src/wp-includes/post.php#L473

With 6.6 https://github.com/WordPress/wordpress-develop/blob/6.6/src/wp-includes/post.php#L476

File: phpunit/class-wp-rest-global-styles-controller-gutenberg-test.php
Comment (ramonjd) [inline phpunit/class-wp-rest-global-styles-controller-gutenberg-test.php:84]: This needs to occur to ensure we're only testing the Gutenberg-registered endpoints...

**Review by ramonjd — COMMENTED** [review]
@Mamaduka looks like things are passing now. 

Basically, it was due to the CI using 6.5.5 so we have to register all missing post args since then.

Thanks @aaronrobertshaw for the inheritance tip and helping to test.

File: lib/rest-api.php
Comment (aaronrobertshaw) [reply lib/rest-api.php:27]: > They should probably be running 6.6

I'm assuming that these tests run on the previous major WP version due to our BC commitment. That said, I did expect we also ran the unit tests on the latest too.

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the persistence here @ramon 💪 

I pushed a small commit adding some [missed updates](https://github.com/WordPress/wordpress-develop/commit/6fc019aa711308d6eaa1e672498b2a789a0c584c#diff-58982cfca0cd603cc3ace01a90ef7f243c7842eb8bf99291667a54f9d94ceeccR53) to the controller so it matches core. Tests still pass locally with this change in place on 6.5.5 and latest. A quick sanity check here would be handy.

---

✅ I could replicate the failing tests locally on 6.5.5 before the last round of fixes
✅ All tests are passing locally and on GitHub across WP versions
✅  Global Styles controller in Gutenberg mostly matches core (route override args excepted)
✅ Manual smoke testing of the site editor looked fine

I'm happy to approve this one now but it's probably wise to get a second opinion from @Mamaduka and co.

**Comment by aaronrobertshaw** [comment]
I've given this PR another look with fresh eyes. I think it's in a good place to be able to merge and unblock https://github.com/WordPress/gutenberg/pull/65071. We can address anything else in a follow-up.

**Comment by Mamaduka** [comment]
Thanks for landing this, @aaronrobertshaw and @ramonjd 🙇

**Pull request merged**
